src/expr/src/relation/mod.rs

```diff
@@ -237,13 +237,11 @@ pub enum MirRelationExpr {
 }
 
 impl MirRelationExpr {
-    /// Reports the schema of the relation.
+    /// Reports the schema of the relation given the schema of the input relations.
     ///
-    /// This method determines the type through recursive traversal of the
-    /// relation expression, drawing from the types of base collections.
-    /// As such, this is not an especially cheap method, and should be used
-    /// judiciously.
-    pub fn typ(&self) -> RelationType {
+    /// It can be used during a post-order traversal.
+    pub fn typ_with_input_types(&self, input_types: &[RelationType]) -> RelationType {
```

Comment:
It would help to have doc'd somewhere what the input types should be. How should I use this correctly, for example, with a `Let` expression? It doesn't need to be a full list of all variants, but .. I chose the one I wouldn't be sure of for example, but avoiding errors of mis-ordered input types would be helpful!